python-packages/smithy-core/smithy_core/serializers.py

```diff
@@ -0,0 +1,335 @@
+        :param value: The double value to write.
+        """
+        self.write_float(schema, value)
+
+    def write_big_integer(self, schema: "Schema", value: int) -> None:
+        """Write a big integer (arbirtrarily large integer) value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The big integer value to write.
+        """
+        self.write_integer(schema, value)
+
+    def write_big_decimal(self, schema: "Schema", value: Decimal) -> None:
+        """Write a big decimal (arbitrarily large float) value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The big decimal value to write.
+        """
+        ...
+
+    def write_string(self, schema: "Schema", value: str) -> None:
+        """Write a string value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The string value to write.
+        """
+        ...
+
+    def write_blob(self, schema: "Schema", value: bytes) -> None:
+        """Write a blob value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The blob value to write.
+        """
+        ...
+
+    def write_timestamp(self, schema: "Schema", value: datetime.datetime) -> None:
+        """Write a timestamp value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The timestamp value to write.
+        """
+        ...
+
+    def write_document(self, schema: "Schema", value: "Document") -> None:
+        """Write a document value to the output.
+
+        :param schema: The shape's schema.
+        :param value: The document value to write.
+        """
+        ...
+
+    def flush(self) -> None:
+        """Flush the underlying data."""
+
+
+@runtime_checkable
+class MapSerializer(Protocol):
+    """Protocol for serializing maps.
+
+    These are responsible for writing any data needed between keys and values as well as
+    any data needed between entries.
+    """
+
+    def entry(self, key: str, value_writer: Callable[[ShapeSerializer], None]):
+        """Write a map entry.
+
+        :param key: The entry's key.
+        :param value_writer: A callable that accepts a shape serializer to write values.
+        """
+        ...
+
+
+class InterceptingSerializer(ShapeSerializer, metaclass=ABCMeta):
+    """A shape serializer capable of injecting data before writing.
+
+    This can, for example, be used to add in structure member keys, commas between
+    structure members, or commas between lists.
+    """
+
+    @abstractmethod
+    def before(self, schema: "Schema") -> ShapeSerializer: ...
+
+    @abstractmethod
+    def after(self, schema: "Schema") -> None: ...
+
+    @contextmanager
+    def begin_struct(self, schema: "Schema") -> Iterator[ShapeSerializer]:
+        delegate = self.before(schema)
+
+        try:
+            with delegate.begin_struct(schema) as s:
+                yield s
+        finally:
+            self.after(schema)
```

Comment:
I'm dusting off my Python here... but does this work like finally in Java where it's run whether there's an error or not? If it is run just before throwing whatever's caught in the try statement, is that what you really want? I think a lot of times these after statements would do stuff like write a comma or something to an output buffer. Do you want to do that when there's an error?

Reply:
Good call. Switched to try/except/else instead.